### ISSUE TITLE
SSL Connection - Windows 10 - GSKit Error 8

### ISSUE THREAD

**Issue by Withnail73** [ibmdb/node-ibm_db#448]
Hi there... recently moved from nonSSL to SSL and am getting the following error when attempting to connect:
```
[IBM][CLI Driver] SQL1109N  The command was not processed because the database manager failed to load the following DLL: "GSKit Error: 8".  SQLSTATE=42724
```
Running on v2.1.0.  The arm file is in the correct directory as far as I can tell and I know it's valid as this is not an issue for colleagues on Windows 7.  Wondering if it was a GSKit conflict, so I added an env PATH  variable  pointing to the gsk8 files within Program Files (x86) but no joy.

```
"DRIVER={DB2};PROTOCOL=TCPIP;DATABASE=xxx;UID=xxxx;PWD=xxxx;HOSTNAME=xxxx;Security=SSL;PORT=50001;SSLServerCertificate=correctpath;QueryTimeout=60;"
```

Any tips on how to resolve would be appreciated.  Thanks!

**Comment by bimalkjha** [comment]
@Withnail73 Add full path of .../ibm_db/clidriver/lib and .../ibm_db/clidriver/lib/icc at start of PATH and LIB env variables. Thanks.

**Comment by Withnail73** [comment]
Thanks, Bimal... the ibm_db install is different to what you have above...the closest I have is:

`...\node_modules\ibm_db\installer\clidriver\lib`
`...\node_modules\ibm_db\installer\clidriver\bin\icc64`

I added these to both User and System PATH and LIB but no luck.  The ibm_db gsk files are in the icc64 folder.  :(

**Comment by bimalkjha** [comment]
@Withnail73 As per https://www.ibm.com/support/knowledgecenter/en/SSAL2T_7.1.0/com.ibm.cics.tx.doc/reference/r_gskit_error_codes.html the GSKit Error 8 means : `GSK_ERROR_CERT_VALIDATION - Certificate validation error.`
It means the certificate file has some issue and validation fails. Could you try getting another certificate file and try? Make sure your certificate file is .arm file. Thanks.

**Comment by Withnail73** [comment]
Yeah, it is in the file - but I wonder if it needs the full chain.  That is not present.  It's odd that it works for everyone else in Windows 7, though.  Maybe I need to switch back to a keystore rather than `SSLServerCertificate`

**Comment by bimalkjha** [comment]
@Withnail73 Yes, you can create keysotre db using GSKit and then try. Full chain is required. Thanks.

**Comment by Withnail73** [comment]
Ok, @bimaljha, I used GSKit to create a new kdb and added all the certs as well as the IBM root one.  Node now throws the 408 error:
```
408 - GSK_ERROR_BAD_KEYFILE_PASSWORD
The specified key file password is incorrect. The key file could not be used. The key file may also be corrupt.
```
There's nothing wrong with the sth file.....as I can use `gsk8capicmd -keydb -expiry -db dmt.kdb -stashed` to open it from a cmd line.  I can only assume it is a permissions issue on the file.  But I granted every user Full Control but still no luck.

**Comment by Withnail73** [comment]
Update - I have another colleague (the only other on Windows 10) who also has the problem. 
```
[IBM][CLI Driver] SQL1109N  The command was not processed because the database manager failed to load the following DLL: "GSKit Error: 8".  SQLSTATE=42724
```
The difference is she doesn't have gsk8 installed as a separate entity - this would appear to rule out any GSK conflicts.

The fact the error message does not explicitly mention any missing DLLs means we can rule out any env PATH issues, too, I suppose.

This leaves the certificate and arm file.  If it couldn't open the file we'd get a 202 or a 408 - so perhaps the  `GSK_ERROR_CERT_VALIDATION - Certificate validation error.` is real - but why would it work on Win7/Linux but not Win10?

Do you understand how GSKit goes about its business?  How does it validate the cert?

**Comment by bimalkjha** [comment]
@Withnail73 .arm certificate files are target database server specific. i.e. if you have generated .arm file from one host, you can use this certificate to connect to that host only. Could you confirm win7 and win10 clients are connecting to same host/db? Thanks.

**Comment by Withnail73** [comment]
Yes, @bimalkjha, they are connecting to the same host/db.  Thanks.

**Comment by bimalkjha** [comment]
@Withnail73 In that case, I think you are hitting Key Management Error: `0x00000008 | 8 | GSKKM_ERR_DATABASE_CREATE` . It is trying to create keydb under `.../ibm_db/installer/clidriver/cfg` directory and user may not have permission to create file under this directory due to OS security reasons. Please try to correct the permission of cfg folder. Thanks.

**Comment by bimalkjha** [comment]
@Withnail73 Please check and correct permission for `C:\ProgramData\IBM\DB2\...._clidriver\cfg` folder too and then try. Under `C:\ProgramData\IBM\DB2` there should be a folder with path of ibm_db clidriver. 
It it still do not work, then we need db2 trace files. To collect trace files, you can edit `installer\testODBCConnection.bat` file and replace the `db2cli validate` command with `node yourtestfile.js`. Save it and run to execute your test file and collect traces for it. Then you can zip all the generated 1.* files and send to bjha@rocketsoftware.com. Thanks.

**Comment by Withnail73** [comment]
Okedoke... traces on their way.  Thanks.

**Comment by bimalkjha** [comment]
@Withnail73 We have published ibm_db@2.4.1 which uses clidriver v11.1 on windows. Please use this version and let us know if you still face the connection problem. thanks.

**Comment by Withnail73** [comment]
Game on!  Thanks, @bimalkjha - works now!  Appreciate the help. 👍
